lib/engine-addon.js

```diff
@@ -387,10 +387,76 @@ module.exports = {
         return mergeTrees([
           externalTree,
           engineJSTree,
-          engineCSSTree
+          compiledEngineCSSTree
         ].filter(Boolean), { overwrite: true });
       };
 
+      this.compileLazyEngineStyles = function compileLazyEngineStyles(vendorTree, externalTree) {
+        var vendorCSSTree = buildVendorCSSTree.call(this, vendorTree);
+        var engineStylesOutputDir = 'engines-dist/' + this.name + '/assets/';
+
+        // Get base styles tree.
+        var engineStylesTree = this._treeFor('addon-styles');
+
+        var primaryStyleTree = null;
+        if (engineStylesTree) {
+          var preprocessedEngineStylesTree;
+          // `_addonPreprocessTree` was added in Ember 2.12 to ensure
+          // addons had preprocessTree / postprocessTree invoked properly
+          // this emulates the normal behavior before and after that version
+          if (this._addonPreprocessTree) {
+            preprocessedEngineStylesTree = this._addonPreprocessTree('css', engineStylesTree);
+          } else {
+            // Ember CLI < 2.12 does not support preprocessTree on addons
+            preprocessedEngineStylesTree = engineStylesTree;
+          }
+
+          var processedEngineStylesTree = preprocessCss(preprocessedEngineStylesTree, '/', '/', {
+            outputPaths: { 'addon':  engineStylesOutputDir + 'engine.css' },
+            registry: this.registry
+          });
+
+          // Move styles tree into the correct place.
+          // `**/*.css` all gets merged.
+          primaryStyleTree = concat(processedEngineStylesTree, {
+            allowNone: true,
+            inputFiles: ['**/*.css'],
+            outputFile: engineStylesOutputDir + 'engine.css'
+          });
+        }
+
+        var concatVendorCSSTree = concat(vendorCSSTree, {
+          allowNone: true,
+          inputFiles: ['**/*.css'],
+          outputFile: engineStylesOutputDir + 'engine-vendor.css'
+        });
+
+        var concatMergedVendorCSSTree = mergeTrees([concatVendorCSSTree, externalTree]);
+
+        // So, this is weird, but imports are processed in order.
+        // This gives the chance for somebody to prepend onto the vendor files.
+        var vendorCSSImportTree = buildVendorCSSWithImports.call(this, concatMergedVendorCSSTree);
+
+        var mergedVendorCSSWithImportAndEngineStylesTree = mergeTrees(
```

Comment:
do we need https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/merge-trees.js ?